xbmc/platform/linux/input/LibInputSettings.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "LibInputSettings.h"
+
+#include "FileItem.h"
+#include "LibInputHandler.h"
+#include "ServiceBroker.h"
+#include "settings/lib/Setting.h"
+#include "settings/lib/SettingsManager.h"
+#include "settings/Settings.h"
+#include "utils/log.h"
+#include "utils/XBMCTinyXML.h"
+
+#include <algorithm>
+
+const std::string CLibInputSettings::SETTING_INPUT_LIBINPUTKEYBOARDLAYOUT = "input.libinputkeyboardlayout";
+static std::vector<std::pair<std::string, std::string>> layouts;
+
+CLibInputSettings::CLibInputSettings(CLibInputHandler *handler) :
+  m_libInputHandler(handler)
+{
+  std::set<std::string> settingSet;
+  settingSet.insert(SETTING_INPUT_LIBINPUTKEYBOARDLAYOUT);
+  CServiceBroker::GetSettings()->GetSettingsManager()->RegisterCallback(this, settingSet);
+
+  CServiceBroker::GetSettings()->GetSettingsManager()->RegisterSettingOptionsFiller("libinputkeyboardlayout", SettingOptionsKeyboardLayoutsFiller);
+
+  /* load the keyboard layouts from xkeyboard-config */
+  std::string xkbFile("/usr/share/X11/xkb/rules/base.xml");
+
+  CFileItem file(xkbFile);
+
+  if (!file.Exists())
+  {
+    CLog::Log(LOGWARNING, "CLibInputSettings: unable to load keyboard layouts from non-existing file: %s", xkbFile.c_str());
+    return;
+  }
+
+  CXBMCTinyXML xmlDoc;
+  if (!xmlDoc.LoadFile(xkbFile))
+  {
+    CLog::Log(LOGWARNING, "CLibInputSettings: unable to open: %s", xkbFile.c_str());
+    return;
+  }
+
+  const TiXmlElement* rootElement = xmlDoc.RootElement();
+  if (rootElement == nullptr)
+  {
+    CLog::Log(LOGWARNING, "CLibInputSettings: missing or invalid XML root element in: %s", xkbFile.c_str());
+    return;
+  }
+
+  if (rootElement->ValueStr() != "xkbConfigRegistry")
+  {
+    CLog::Log(LOGWARNING, "CLibInputSettings: unexpected XML root element %s in: %s", rootElement->Value(), xkbFile.c_str());
+    return;
+  }
+
+  const TiXmlElement* layoutElement = rootElement->FirstChildElement("layoutList")->FirstChildElement("layout");
```

Comment:
Nested elements need null checks in TinyXML

Reply:
thanks, fixed